text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,440 @@
+
+Now suppose we wanted to extend our TinyTodo policies with the following:
+
+```
+// Policy 6: No access if not high rank and at location DEF, 
+// or at resource's owner's location
+forbid(
+    principal,
+    action,
+    resource is List
+) unless {
+   principal.joblevel > 6 && principal.location like "DEF*" ||
+   principal.location == resource.owner.location
+};
+```
+
+This policy does *not* validate at level 1. That is because the expression `resource.owner.location` is only valid at level 2: It requires dereferencing `resource` to retrieve the entity in the `owner` attribute, and then requires dereferencing that entity to obtain the contents of its `location` attribute. As a result, if we only provided the `principal` and `resource` entity data, as described above, evaluating policy 6 could produce an evaluation error since data may not be provided for `resource.owner`.
+
+All of the Cedar example policy sets validate with level-based validation, either at level 1 or 2; see the Appendix for details.
+
+## Motivation
+
+Currently, Cedar provides no support for slicing. Users are required to either be inefficient and produce their entire entity store as a slice, or derive their own ad-hoc slicing algorithm. We'd like to provide an out-of-the-box solution that addresses many user's needs. Entity slicing in general is a problem very tied to the specifics of application/deployment. It's likely impossible to have a fully general solution. This RFC is attempting to provide a solution for entity slicing that works for the 80% case, while acknowledging that it won't work for everyone. 
+
+Additionally, there are some application scenarios where the actors writing the policies and the people performing slicing are not the same people, and thus the slicers need to impose some kind of restrictions on policy authors.
+
+This RFC has a close connection to [RFC 74: Entity Slicing using Data Tries](https://github.com/cedar-policy/rfcs/pull/74), which also seeks to provide a sound foundation for entity slicing. We discuss the relationship between them in the Alternatives below; ultimately, we believe they are complementary and should coexist.
+
+## Detailed design
+
+### Entity dereferences
+
+An "Entity Dereference" is any operation that requires directly accessing an entity’s data. The following is an exhaustive list of entity dereferencing operations
+
+1. `e1 in e2` dereferences `e1` (but not `e2`)
+2. `e1.foo` dereferences `e1` iff `e1` is an entity
+3. `e1 has foo` dereferences `e1` iff `e1` is an entity
+
+Note that testing equality *is not* a dereferencing operation. Therefore expressions like `e1 == e2`, `e1.contains(e2)` and `e3 in e2` do not dereference `e1` or `e2`.
```

Comment:
```suggestion
Note that testing equality *is not* a dereferencing operation. Therefore expressions like `e1 == e2`, `e1.contains(e2)` and `e1 in e2` do not dereference `e1` or `e2`.
```

Reply:
Note: `e1 in e2` _does_ dereference `e1`, which is why I wrote `e3` there.

Reply:
My initial reading of this also thought it was a typo.  Is readability better if you combine DOES and DOES NOT deference into the same list, so `e1 in e2` and `e3 in e2` examples are closer together?

Also, does the reflexive `e1 in e1` dereference `e1` or is that equality?